### DISCUSSION TITLE
How to convert the uploaded image to Numpy array?

### DISCUSSION THREAD

**Discussion by pavitrashah** [fastapi/fastapi#7048] [Questions] [answered]
### Example Code To Get an Image

```Python
from fastapi import FastAPI, UploadFile, File, Form

app = FastAPI()


@app.post("/")
def read_root(file: bytes = File(...)):
    return {"Hello": "World"}
```
#### Or

```Python
from fastapi import FastAPI, UploadFile, File, Form

app = FastAPI()


@app.post("/")
def read_root(file: UploadFile = File(...)):
    return {"Hello": "World"}
```

### Environment

* OS: Windows 10
* FastAPI Version == 0.61.2
* Python version: 3.7

### Question

How to convert the uploaded image to Numpy array? So it can be then used in libraries like openCV, tensorflow for Computer Vision or Deep Learning Applications.

### Things I have already tried
```Python
from fastapi import FastAPI, UploadFile, File, Form
from PIL import Image
from io import BytesIO
import numpy as np

app = FastAPI()

def read_imagefile(data) -> Image.Image:
    image = Image.open(BytesIO(data))
    return image

@app.post("/")
async def read_root(file: UploadFile = File(...)):
    image = read_imagefile(await file.read())
    return {"Hello": "World"}
```
#### And

```Python
from fastapi import FastAPI, UploadFile, File, Form
from PIL import Image
from io import BytesIO
import numpy as np

app = FastAPI()

def load_image_into_numpy_array(data):
    return np.array(Image.open(BytesIO(data)))


@app.post("/")
async def read_root(file: UploadFile = File(...)):
    image = load_image_into_numpy_array(await file.read())
    return {"Hello": "World"}
```

But I keep getting errors. And none of the solutions in any other thread works.
Errors: `AttributeError: 'JpegImageFile' object has no attribute 'read'` or
`AttributeError: 'bytes' object has no attribute 'read'` or
`AttributeError: 'numpy.ndarray' object has no attribute 'read'` or
`UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte`

**Comment by includeamin** [comment, marked as answer]
both of the codes are working for me.
``` python
@app.post("/")
async def read_root(file: UploadFile = File(...)):
    image = load_image_into_numpy_array(await file.read())
    print(image.data)
    return {"Hello": "World"}
```
``` sh
curl -X POST "http://localhost:3000/" -H  "accept: application/json" -H  "Content-Type: multipart/form-data" -F "file=@image (31).png;type=image/png"

```
here the output:
![image](https://user-images.githubusercontent.com/11836741/99591360-eaf31600-2a03-11eb-9666-124e5c42bbbd.png)

**Comment by pavitrashah** [comment]
Thanks for the reply. Both the codes also worked out for me apparently I had made a mistake in other part of my code.

**Comment by tiangolo** [comment]
Thanks for the help here @includeamin ! :clap: :bow: 

Thanks for reporting back and closing the issue @pavitrashah  :+1:

**Comment by belito3** [comment]
Probably, we need convert image's color from BGR to RGB

```
def load_image_into_numpy_array(data):
    npimg = np.frombuffer(data, np.uint8)
    frame = cv2.imdecode(npimg, cv2.IMREAD_COLOR)
    cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    return frame
```

**Comment by chaitanyakkumar** [comment]
hi guys. I am working on  a similar code like yours but instead of printing the numpy array I need to return the numpy array. While returning the numpy array I am getting internal server error

**Comment by Noctale** [comment]
[https://pynative.com/python-serialize-numpy-ndarray-into-json/#:~:text=Use%20the%20cls%20kwarg%20of,array%20into%20JSON%20formatted%20data.&text=To%20serialize%20Numpy%20array%20into%20JSON%20we%20need%20to%20convert,using%20a%20tolist()%20function.](url)

this work for me to return numpy (but it serialized)

**Comment by maazbin** [comment]
Thanks for the code. It worked and solved my problem.

**Comment by michaeleliot** [comment]
Had a lot of issues with this. Here is my final code:

Image Conversion File
``` python
def file_to_image(file):
  image = Image.open(BytesIO(file)).convert("RGB")
  return image
```
FastApi File
``` python
...
@app.post("/file")
async def create_upload_file(file: UploadFile):
    raw_file = await file.read()
    image = file_to_image(raw_file)
...
```
Hope this helps!
